From 99a988557509536fe57cd9ad8dd6b1f489e223c8 Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:36:28 -0400
Subject: [ethernet] i40e: delete netdev after deleting napi and vectors

Message-id: <1409658042-22689-28-git-send-email-sassmann@redhat.com>
Patchwork-id: 90222
O-Subject: [RHEL7.1 PATCH 027/281] i40e: delete netdev after deleting napi and vectors
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Shannon Nelson <shannon.nelson@intel.com>

We've been deleting the netdev before getting around to deleting the napi
structs.  Unfortunately, we then didn't delete the napi structs because we
have a check for netdev, thus we were leaving garbage around in the system.

Change-ID: Ife540176f6c9f801147495b3f2d2ac2e61ddcc58
Signed-off-by: Shannon Nelson <shannon.nelson@intel.com>
Signed-off-by: Catherine Sullivan <catherine.sullivan@intel.com>
Tested-by: Kavindya Deegala <kavindya.s.deegala@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit a48665970962a9b50aa81722ca4e943fcfdc6699)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_main.c b/drivers/net/ethernet/intel/i40e/i40e_main.c
index 43d391b..a3f122e 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_main.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_main.c
@@ -6806,8 +6806,6 @@ int i40e_vsi_release(struct i40e_vsi *vsi)
    if (vsi->netdev) {
     /* results in a call to i40e_close() */
     unregister_netdev(vsi->netdev);
-    free_netdev(vsi->netdev);
-    vsi->netdev = NULL;
    }
   } else {
    if (!test_and_set_bit(__I40E_DOWN, &vsi->state))
@@ -6826,6 +6824,10 @@ int i40e_vsi_release(struct i40e_vsi *vsi)
 
  i40e_vsi_delete(vsi);
  i40e_vsi_free_q_vectors(vsi);
+ if (vsi->netdev) {
+  free_netdev(vsi->netdev);
+  vsi->netdev = NULL;
+ }
  i40e_vsi_clear_rings(vsi);
  i40e_vsi_clear(vsi);
 
-- 
1.7.1